plugins/state_history_plugin/include/eosio/state_history_plugin/session.hpp

```diff
@@ -250,16 +250,16 @@ class session final : public session_base {
                      block_to_send->blocks_result_base.this_block  = {self.current_blocks_request.start_block_num, *this_block_id};
                      if(const std::optional<chain::block_id_type> last_block_id = self.get_block_id(self.next_block_cursor - 1))
                         block_to_send->blocks_result_base.prev_block = {self.next_block_cursor - 1, *last_block_id};
-                     if(chain::signed_block_ptr sbp = get_block(self.next_block_cursor); sbp && self.current_blocks_request.fetch_block)
+                     if(chain::signed_block_ptr sbp = get_block(*this_block_id); sbp && self.current_blocks_request.fetch_block)
                          block_to_send->blocks_result_base.block = fc::raw::pack(*sbp);
                      if(self.current_blocks_request.fetch_traces && self.trace_log)
                         block_to_send->trace_entry = self.trace_log->get_entry(self.next_block_cursor);
                      if(self.current_blocks_request.fetch_deltas && self.chain_state_log)
                         block_to_send->state_entry = self.chain_state_log->get_entry(self.next_block_cursor);
                      if(block_to_send->is_v1_request && *self.current_blocks_request_v1_finality && self.finality_data_log)
                         block_to_send->finality_entry = self.finality_data_log->get_entry(self.next_block_cursor);
+                     ++self.next_block_cursor;
```

Comment:
It seems like this change may cause ship to enter a tight busy loop, of sorts, and spam the client forever if we ever enter this failure condition. Because `next_block_cursor` will be stuck forever at a "bad block" (one we can't look up the id to) and after sending this nearly empty block result to the client we'll just reconsider the "bad block" once again immediately once the async_send compeltes. Repeat.

Possibly the discussion over in https://github.com/AntelopeIO/spring/pull/365/files#r1679469382 is right that we could consider asserting `get_block_id(self.next_block_cursor)` succeeds and kill off the connection when it fails. I am not entirely sure off hand what user visible changes this will cause though. It might mean asking for a blocknum before knowledge of all logs (like if you had started from snapshot w/o any previous logs) will now kill the connection instead of streaming "empty" entries until the first populated entry.

Maybe you can say that sort of request is an erroneous request though (i.e. should have done a get_status_request first to see the valid ranges that ship knows about; though that status request doesn't have the block log range for that corner case of no ship logs enabled)

Reply:
Seems we should continue to allow requests outside of valid range (move the `++self.next_block_cursor` outside the `if`). Seems like the best approach is to stream "empty" entries until the first populated entry.

Reply:
I was worry about an infinite loop on a bad block ID.

Reply:
Add a comment in the code about the empty entries.

Reply:
The main problem with those empty entries is they provide no information to the client what block number they correspond to. To an unaware client things might just look broken. Potentially we could change this with the v1 blocks reply and include some always-available bnum; though I'm not convinced it's worth it. (Even though it's racy if pruning/splitting is enabled) well behaved clients that can't assume log availability from genesis really ought to query status first for the minimum starting block to stream from (which, granted, is only available for trace, state, and finality -- not blog -- that too could potentially be changed with v1 status reply but again not sure it's worth it)

Reply:
I think that is something to consider for a future PR. In practice what we have now should suffice I think.